firedrake/output.py

```diff
@@ -651,3 +756,366 @@ def write(self, *functions, **kwargs):
+
+    def load_function_space(self, name, mesh_name, distribution_parameters={}):
+        # Deal with `MixedFunctionSpace`s
+        if PREFIX + "_mixed_function_spaces" in self.h5pyfile["/topologies"]:
+            base_base_path = os.path.join("/topologies",
+                                          PREFIX + "_mixed_function_spaces")
+            if name in self.h5pyfile[base_base_path]:
+                base_path = os.path.join(base_base_path, name)
+                n = self.get_attr(base_path, PREFIX + "_num_sub_spaces")
+                Vsub_list = []
+                for i in range(n):
+                    path = os.path.join(base_path, str(i))
+                    Vsub_mesh_name = self.get_attr(path, PREFIX + "_mesh")
+                    Vsub_name = self.get_attr(path, PREFIX + "_function_space")
+                    Vsub = self.load_function_space(Vsub_name,
+                                                    Vsub_mesh_name,
+                                                    distribution_parameters=distribution_parameters)
+                    Vsub_list.append(Vsub)
+                V = functools.reduce(lambda a, b: a * b, Vsub_list)
+                V.name = name
+                return V
+        dist_key = hash(frozenset(distribution_parameters.items()))
+        mesh_key = (mesh_name, dist_key)
+        fs_key = (name, ) + mesh_key
+        if fs_key in self._function_spaces:
+            return self._function_spaces[fs_key]
+        # Load mesh
+        mesh = self.load_mesh(mesh_name, distribution_parameters=distribution_parameters)
+        mesh.init()
+        # Load element
+        sectiondm_name = self.find_sectiondm_name(mesh_name, name)
+        path = os.path.join("/topologies",
+                            mesh.topology_dm.getName(),
+                            "dms",
+                            sectiondm_name,
+                            PREFIX + "_function_spaces",
+                            name)
+        element_b = self.get_attr(path, PREFIX + "_ufl_element").tobytes()
+        element = pickle.loads(element_b)
+        # Load shared data
+        tmesh = mesh.topology
+        dm = tmesh.topology_dm
+        sectiondm = PETSc.DMShell().create(comm=dm.comm)
+        sectiondm.setName(sectiondm_name)
+        sectiondm.setPointSF(dm.getPointSF())
+        section = PETSc.Section().create(comm=dm.comm)
+        section.setPermutation(tmesh._plex_renumbering)
+        sectiondm.setSection(section)
+        sfXB = self._meshes[mesh_key]["sfXB"]
+        sfXC = self._meshes[mesh_key].setdefault("sfXC",
+                                                 sfXB.compose(tmesh.sfBC) if tmesh.sfBC else sfXB)
+        gsf, lsf = dm.sectionLoad(self.viewer, sectiondm, sfXC)
+        sd_key = get_shared_data_key_for_checkpointing(tmesh, element)
+        _ = get_global_numbering(tmesh, sd_key, global_numbering=sectiondm.getSection())
+        # Construct function space
+        V = FunctionSpace(mesh, element, name=name)
```

Comment:
Can we get around this by just constructing the functionspace on the loaded (redistributed?) mesh and then computing the relevant sf that pushes forward? AIUI, `dm.sectionLoad` populates the section inside the sectiondm by pushing forward dofs in the on disk section through sfXC. But we already know how many dofs will be on each point, and we just need to line them up (or is this harder?)

Reply:
With #2154, I've now made it clear that we use the same cache key generator here and in `functionspacedata`.

`sfXC` is an SF that pushes the global number slab (`X = [0, N)`) to the good plex (redistributed and overlapped as we have already called `mesh.init()`). `DMPlexSectionLoad()` does the following:
- Load raw section (`sectionA`) with `PetscSectionLoad()`
- Construct `sfAX` that pushes points in `sectionA` to the slab `X`
- Compose `sfAC = SFCompose(sfAX, sfXC)`
- `PetscSFDistributeSection(sfAC, sectionA, &remoteOffsets, sectionC)` (`sectionC` is the section on the good plex)
- Use `remoteOffsets` to construct `globalDataSF` (or `localDataSF`) that directly pushes on-disk vector data to global vectors (or local vectors) defined on `sectionC`

As you noted, the resulting `sectionC` would be identical to what we would construct using the UFL element (because we happen to know the definite rule to go from the UFL element to the section), but it is also necessary to  construct `globalDataSF` and `localDataSF` from `remoteOffsets`, which knows the on-disk array offsets, which looks hard to do outside PETSc.

I wanted to call `get_global_numbering` here to avoid reconstruction of `global_numbering` as `dm.sectionLoad` constructs it anyway. I tried to treat the case where `global_numbering` has already been in the cache. Does it look right?